File: windows-client/ComponentVault.WinUI/Localization/CategoryFilterSemantics.cs
using System.Globalization;
using ComponentVault.WinUI.Services.Catalog;

namespace ComponentVault.WinUI.Localization;

internal static class CategoryFilterSemantics
{
    internal static string Key(string? category) => LcscPublicCatalog.NormalizeCategory(category?.Trim() ?? string.Empty);

    internal static IReadOnlyList<string> Options(IEnumerable<string> categories) => categories
        .Where(category => !string.IsNullOrWhiteSpace(category))
        .Select(category => category.Trim())
        .GroupBy(Key, StringComparer.OrdinalIgnoreCase)
        .Select(group => group.FirstOrDefault(category => !Key(category).Equals(category, StringComparison.Ordinal)) ?? group.First())
        .OrderBy(category => category, StringComparer.OrdinalIgnoreCase)
        .ToArray();

    internal static bool SameCategory(string rawCategory, string selectedKey) =>
        Key(rawCategory).Equals(Key(selectedKey), StringComparison.OrdinalIgnoreCase);

    internal static bool MatchesSearch(string rawCategory, string query)
    {
        var needle = query.Trim();
        return needle.Length == 0
            || rawCategory.Contains(needle, StringComparison.OrdinalIgnoreCase)
            || Key(rawCategory).Contains(needle, StringComparison.OrdinalIgnoreCase)
            || CategoryDisplay.Localize(rawCategory, CultureInfo.GetCultureInfo("zh-CN")).Contains(needle, StringComparison.OrdinalIgnoreCase);
    }
}

File: android-client/app/src/main/java/com/componentvault/android/ui/screen/CategoryFilterSemantics.kt
package com.componentvault.android.ui.screen

import com.componentvault.android.data.OfficialCategoryNormalizer
import java.util.Locale

internal object CategoryFilterSemantics {
    fun key(category: String): String = OfficialCategoryNormalizer.normalize(category)

    fun options(categories: Iterable<String>): List<String> = categories
        .map(String::trim)
        .filter(String::isNotBlank)
        .groupBy { key(it).lowercase(Locale.ROOT) }
        .values
        .map { group -> group.firstOrNull { key(it) != it } ?: group.first() }
        .sortedBy { it.lowercase(Locale.ROOT) }

    fun sameCategory(rawCategory: String, selectedKey: String): Boolean =
        key(rawCategory).equals(key(selectedKey), ignoreCase = true)

    fun matchesSearch(rawCategory: String, query: String): Boolean {
        val needle = query.trim()
        if (needle.isEmpty()) return true
        return rawCategory.contains(needle, ignoreCase = true) ||
            key(rawCategory).contains(needle, ignoreCase = true) ||
            localizedCategoryLabel(rawCategory, Locale.SIMPLIFIED_CHINESE).contains(needle, ignoreCase = true)
    }
}
